New JS interop (#27083)
dotnet/AspNetCore.Docs: aspnetcore/blazor/javascript-interoperability/import-export-interop.md

Blazor provides its own JS interop mechanism based on the <xref:Microsoft.JSInterop.IJSRuntime> interface, which is uniformly supported across Blazor hosting models and described in the following articles:

* <xref:blazor/js-interop/call-javascript-from-dotnet>
* <xref:blazor/js-interop/call-dotnet-from-javascript>

<xref:Microsoft.JSInterop.IJSRuntime> enables library authors to build JS interop libraries that can be shared across the Blazor ecosystem and remains the recommend approach for JS interop in Blazor. However, this article describes an alternative JS interop approach available for the first time with the release of .NET 7. The approaches described in this article should be used to replace the obsolete unmarshalled JS interop API of prior Blazor framework releases.

## Obsolete JavaScript interop API

Unmarshalled JS interop using <xref:Microsoft.JSInterop.IJSUnmarshalledRuntime> API is obsolete in ASP.NET Core 7.0. Follow the guidance in this article to replace the obsolete API.

## Prerequisites

[!INCLUDE[](~/includes/7.0-SDK.md)]

## Call JavaScript from .NET

This section explains how to call JS functions from .NET.

> [!NOTE]

> > Add the <xref:Microsoft.Build.Tasks.Csc.AllowUnsafeBlocks> property to the app's project file. Setting this property to `true` permits the code generator in the Roslyn compiler to use pointers and isn't a security concern.

Reviewer: I would not even say `and isn't a security concern.`

            message = Interop.GetWelcomeMessage();
        }
    }
}
```

> [!NOTE]
> The conditional check for <xref:System.OperatingSystem.IsBrowser%2A?displayProperty=nameWithType> in the preceding example ensures that the code is only called in Blazor WebAssembly apps running on the client. This is important for library code deployed as NuGet packages that might be referenced by developers of Blazor Server apps, where the preceding code can't execute. If you know for sure that the code is only implemented in Blazor WebAssembly apps, you can remove the check for <xref:System.OperatingSystem.IsBrowser%2A?displayProperty=nameWithType>.

If the JS module interacts directly with the component's rendered UI, import the module in [`OnAfterRenderAsync`](xref:blazor/components/lifecycle#after-component-render-onafterrenderasync) and call the imported JS function:

```csharp
protected override async Task OnAfterRenderAsync(bool firstRender)
{
    if (OperatingSystem.IsBrowser() && firstRender)
    {
        await JSHost.ImportAsync("Interop", "../Pages/CallJavaScript.razor.js");

        Interop.DoSomething();
    }
}
```

> If the code holds a reference to the <xref:System.Runtime.InteropServices.JavaScript.JSObject> returned by `JSHost.ImportAsync`, be sure to call <xref:System.Runtime.InteropServices.JavaScript.JSObject.Dispose%2A?displayProperty=nameWithType> when the component is disposed:

Reviewer: This is interesting comment. The `JSObject` returned by the `ImportAsync` is indeed disposable proxy. But even if you dispose it, the `JSImport` infrastructure will keep the JS side ES6 module alive and callable.

In the following example, a JS function named `getMessage` is exported from a collocated JS file that returns a welcome message string, "Hello from Blazor!" in Portuguese:

`Pages/CallJavaScript.razor.js`:

```javascript
export function getMessage() {
  return 'Olá do Blazor!';
}
```

## Call .NET from JavaScript

This section explains how to call .NET methods from JS.

> [!NOTE]
> Add the <xref:Microsoft.Build.Tasks.Csc.AllowUnsafeBlocks> property to the app's project file. Setting this property to `true` permits the code generator in the Roslyn compiler to use pointers and isn't a security concern.

To export a .NET method so that it can be called from JS, use the `[JSExport]` attribute.

In the following example:

* `DotNetInteropCall` is a .NET method with the `[JSExport]` attribute that returns a welcome message string, "Hello from Blazor!" in Portuguese.

> * `SetWelcomeMessage` calls a JS function named `setmessage`. The JS function calls into .NET to receive the welcome string from `DotNetInteropCall` and displays the welcome string in the UI.

Reviewer: upercase `setMessage`

`Interop.cs`:

```csharp
using System.Runtime.InteropServices.JavaScript;
using System.Runtime.Versioning;

[SupportedOSPlatform("browser")]
public partial class Interop
{
    [JSExport]

>     internal static string DotNetInteropCall()

Reviewer: Rename to something more descriptive like `GetMessageFromDotnet`

In the following example, a JS function named `setMessage` is exported from a collocated JS file.

The `setMessage` method:

* Calls `globalThis.getDotnetRuntime(0)` to expose the WebAssembly .NET runtime instance for calling exported .NET methods.
* Obtains the app assembly's JS exports. The name of the app's assembly in the following example is `BlazorSample`.
* Calls the `Interop.DotNetInteropCall()` method from the exports (`exports`). The returned value, which is the welcome string, is assigned to the preceding `<span>`'s inner text.

```javascript
export async function setMessage() {
  const { getAssemblyExports } = await globalThis.getDotnetRuntime(0);
  var exports = await getAssemblyExports("BlazorSample.dll");

  document.getElementById("result").innerText = 

>     exports.Interop.DotNetInteropCall();

Reviewer: Consider adding C# namespace to razor file and here as well.

}
```

## Additional example

For an additional example of the JS interop techniques described in this article, see the *Blazor WASM demo of new JSInterop* sample app:

* [Reference source (`pavelsavara/blazor-wasm-hands-pose` GitHub repository)](https://github.com/pavelsavara/blazor-wasm-hands-pose)
* [Live demonstration](https://pavelsavara.github.io/blazor-wasm-hands-pose/)

> [!NOTE]

> > The [`pavelsavara/blazor-wasm-hands-pose` GitHub repository](https://github.com/SteveSandersonMS/BlazorOnGitHubPages) isn't owned, maintained, or supported by the .NET foundation or Microsoft.

Reviewer: Wrong like to Steve's demo

> The module name in the `[JSImport]` attribute and the call to load the module in the component with `JSHost.ImportAsync` syntax should match. The module name must also be unique in the app. When authoring a library for deployment in a NuGet package, we recommend using the NuGet package namespace as a prefix in module names. In the following example, the module name reflects the `Contoso.InteropServices.JavaScript` package and a folder of user message interop classes (`UserMessages`):

Reviewer: I don't understand "syntax" in this sentence.

Author: Ok ... 👍 ... I'll fix that up on the next commit.

* The imported `setMessage` JS function is called by `SetWelcomeMessage`.
* The returned welcome message is displayed by `setMessage` in the UI via the `message` field.

> [!IMPORTANT]
> In this section's example, JS interop is used to mutate a DOM element *purely for demonstration purposes* after the component is rendered in [`OnAfterRender`](xref:blazor/components/lifecycle#after-component-render-onafterrenderasync). Typically, you should only mutate the DOM with JS when the object doesn't interact with Blazor. The approach shown in this section is similar to cases where a third-party JS library is used in a Razor component, where the component interacts with the JS library via JS interop, the third-party JS library interacts with part of the DOM, and Blazor isn't involved directly with the DOM updates to that part of the DOM. For more information, see <xref:blazor/js-interop/index#interaction-with-the-document-object-model-dom>.

`Pages/CallDotNet1.razor`:

```razor
@page "/call-dotnet-1"
@using System.Runtime.InteropServices.JavaScript

<h1>

>     .NET JS <code>[JSImport]</code>/<code>[JSExport]</code> Interop 1 (Call .NET)

Reviewer: "Interop 2" maybe ?

Reviewer: well, I don't know what that number means here. There is also `CallDotNet1`

Author: This is numbering the example, not the type of interop. I could add the word "Example" in there.

Author: These are named to not conflict 💥 either in a dev's local test app or in the snippet sample app. It's explained to readers here ...

https://learn.microsoft.com/en-us/aspnet/core/blazor/fundamentals/?view=aspnetcore-6.0#razor-components

Reviewer: I see the #2 examples now. I guess it's OK. I was just thinking what is the scope of the numbering (as a reader)

Reviewer: Could we do header for start of example #2 ?

Author: "Example" stuck in there is a good idea. I do that quite a bit around the docs.

Author: > Could we do header for start of example 2 ?

It's in a separate section with the heading, "*Use of a single JavaScript module across components*."

`Pages/CallDotNet1.razor.js`:

```javascript
export async function setMessage() {
  const { getAssemblyExports } = await globalThis.getDotnetRuntime(0);
  var exports = await getAssemblyExports("BlazorSample.dll");

  document.getElementById("result").innerText = 
    exports.BlazorSample.Pages.CallDotNet1.GetMessageFromDotnet();
}
```

> [!NOTE]
> Calling `getAssemblyExports` to obtain the exports can occur in a [JavaScript initializer](xref:blazor/js-interop/index#javascript-initializers) for availability across the app.

## Multiple module import calls

After a JS module is loaded, the module's JS functions are available to the app's components and classes as long as the app is running in the browser window or tab without the user manually reloading the app. `JSHost.ImportAsync` can be called multiple times on the same module in the following cases without a significant performance penalty:

> 

Reviewer: "That could happen when:"

* `SetWelcomeMessage` calls a JS function named `setMessage`. The JS function calls into .NET to receive the welcome string from `GetMessageFromDotnet` and displays the welcome string in the UI.

`Interop2.cs`:

```csharp
using System.Runtime.InteropServices.JavaScript;
using System.Runtime.Versioning;

[SupportedOSPlatform("browser")]
public partial class Interop2
{
    [JSExport]
    internal static string GetMessageFromDotnet()
    {
        return "Olá do Blazor!";
    }

    [JSImport("setMessage", "Interop2")]
    internal static partial void SetWelcomeMessage();
}
```

> [!NOTE]
> For demonstration purposes, the `Interop2` class doesn't include a namespace declaration. In production apps, using namespaces for JS interop C# classes is recommended.

The following `CallDotNet1` component calls JS that directly interacts with the DOM to render the welcome message:

* The `Interop2` [JS module](xref:blazor/js-interop/index#javascript-isolation-in-javascript-modules) is imported asychronously from the collocated JS file for this component.

> * The imported `setMessage` JS function is called with `Interop.SetWelcomeMessage()`.

Author: Ah! Yes ... I missed two updates. The next commit will get those fixed up.

---
title: .NET JavaScript `[JSImport]`/`[JSExport]` interop with ASP.NET Core Blazor WebAssembly
author: guardrex
description: Learn how to interact with JavaScript in Blazor WebAssembly apps using .NET JavaScript `[JSImport]`/`[JSExport]` interop.
monikerRange: '= aspnetcore-7.0'
ms.author: riande
ms.custom: mvc
ms.date: 09/28/2022
uid: blazor/js-interop/import-export-interop
---
# .NET JavaScript `[JSImport]`/`[JSExport]` interop with ASP.NET Core Blazor

This article explains how to interact with JavaScript (JS) in Blazor WebAssembly apps using .NET JS `[JSImport]`/`[JSExport]` interop API released with .NET 7.

Blazor provides its own JS interop mechanism based on the <xref:Microsoft.JSInterop.IJSRuntime> interface, which is uniformly supported across Blazor hosting models and described in the following articles:

* <xref:blazor/js-interop/call-javascript-from-dotnet>
* <xref:blazor/js-interop/call-dotnet-from-javascript>

> <xref:Microsoft.JSInterop.IJSRuntime> enables library authors to build JS interop libraries that can be shared across the Blazor ecosystem and remains the recommend approach for JS interop in Blazor.

Reviewer 2: ```suggestion
<xref:Microsoft.JSInterop.IJSRuntime> enables library authors to build JS interop libraries that can be shared across the Blazor ecosystem and remains the recommended approach for JS interop in Blazor.
```

> This article describes an alternative JS interop approach available for the first time with the release of .NET 7. The approaches described in this article can be used in place of existing <xref:Microsoft.JSInterop.IJSRuntime> approaches and should be used to replace obsolete unmarshalled JS interop API when migrating to .NET 7.

Reviewer 2: Somewhere we need to provide guidance on when to use the JSImport/JSExport model vs IJSRuntime for both app and library authors. How about this?:

```suggestion
This article describes an alternative JS interop approach specific to WebAssembly-based apps available for the first time with the release of .NET 7. These approaches are appropriate when you only expect to run on WebAssembly and not in the other Blazor hosting models. Library authors can use these approaches to optimize JS interop on WebAssembly by checking at runtime if the app is running on WebAssembly (OperatingSystem.IsBrowser). The approaches described in this article should be used to replace obsolete unmarshalled JS interop API when migrating to .NET 7.
```

## Obsolete JavaScript interop API

Unmarshalled JS interop using <xref:Microsoft.JSInterop.IJSUnmarshalledRuntime> API is obsolete in ASP.NET Core 7.0. Follow the guidance in this article to replace the obsolete API.

## Prerequisites

[!INCLUDE[](~/includes/7.0-SDK.md)]

> ## Cache busting during development

Author: IIRC, the JS FNs loaded by the module are getting cached here across VS debugging sessions ... across VS restarts of debugging ... i.e., there seemed like there was some browser caching going on. I don't know/recall the exact conditions. I just know that on several occasions I had to force reload the app to get updated JS FNs to load (and/or I had the dev tools console open with static caching disabled). I don't recall the exact scenarios, but I can work them out if needed. I was under the impression from chatting with @pavelsavara that he saw it under certain conditions and may know when/where/why it happens.

> ## Cache busting during development

Reviewer 2: @guardrex Were you seeing these caching issues in a Blazor app? Or was this with a Blazor WebAssembly App? Can you help me reproduce the behavior you saw?

> ## Cache busting during development

Author: Yeah ... it's the exact example on the topic in a Blazor WebAssembly app from the Blazor WebAssembly template.

I've updated my test app at GH (Pavel was using it earlier) ...

https://github.com/guardrex/BlazorWASM70UnmarshalledJS/

Ignore that name ... that app was set up prior to my knowing anything about the new API. Also, ignore the `OperatingSystem.IsBrowser()` in the app. Pavel and I ultimately just call that out in the text for the topic's example. We don't show it in the article's component examples.

Try the `CallJavaScript1` component example (it's on the `NavMenu`, so you can just run the app) ... change the return string in the JS file for it (`Pages/CallJavaScript1.razor.js`) while debugging. You can navigate away to another component and back ... no dice. I can't get an update of the JS client-side without the hard reload on one side or the other.

> ## Cache busting during development

Reviewer 2: @pavelsavara When JSHost.ImportAsync gets invoked a 2nd time does it do its own caching of the imported module? When I run this app I see JSHost.ImportAsync get called but I don't see any network activity at all.

> ## Cache busting during development

Reviewer: There is browser caching it for the same URL and we are also caching it by assigned module name in runtime.

> ## Cache busting during development

Reviewer 2: I removed the cache busting logic from the component, and now I'm seeing this:

![image](https://user-images.githubusercontent.com/1874516/194620006-e8b62a82-3639-4e90-a7fc-3445d7c946cb.png)

Now I'm trying to figure out where the cache busting is coming from in the 2nd request after I refresh the page. @pavelsavara I assume that's not being done by JSHost? I suspect it's the something in static web assets.

> ## Cache busting during development

Reviewer 2: I confirmed with @javiercn that it isn't static web assets doing the cache busting in this case. @pavelsavara Is it JSHost?

> ## Cache busting during development

Reviewer: the JSHost is not doing it `?Cache=`. Did your code have `bustCache` or `Cache` query ? Did just hot reload failed you here ?

When implementing JS interop, browser caching of JS files during development can interfere with writing and testing JS interop code. To force a browser to reload JS files (cache busting) during development, we recommend using browser [developer tools](https://developer.mozilla.org/docs/Glossary/Developer_Tools) with static asset caching disabled. For more information, access the documentation for the developer tools associated with your browser:

* [Chrome DevTools](https://developer.chrome.com/docs/devtools/)
* [Firefox Developer Tools](https://developer.mozilla.org/docs/Tools)
* [Microsoft Edge Developer Tools overview](/microsoft-edge/devtools-guide-chromium/)

> ## Enable unsafe blocks

Reviewer 2: @pavelsavara @lewing @SteveSandersonMS  Is there a way we can restrict this to just the code generator so it doesn't impact the rest of the user's app?

Reviewer: No AFAIK

Reviewer: They could separate the interop code to separate assembly.

Reviewer: But that's probably overkill.

Reviewer 2: Let's discuss offline if there's something we could do to improve this. For example could we add a way to enable unsafe blocks for only the generated code? But we don't need to block this PR on this discussion.

monikerRange: '= aspnetcore-7.0'
ms.author: riande
ms.custom: mvc
ms.date: 10/10/2022
uid: blazor/js-interop/import-export-interop
---
# .NET JavaScript `[JSImport]`/`[JSExport]` interop with ASP.NET Core Blazor

This article explains how to interact with JavaScript (JS) in Blazor WebAssembly apps using .NET JS `[JSImport]`/`[JSExport]` interop API released with .NET 7.

> [!NOTE]
> This article focuses on JS interop in Blazor WebAssembly apps. For guidance on using .NET JS `[JSImport]`/`[JSExport]` interop API in JavaScript apps, see <xref:client-side/import-export-interop>.

Blazor provides its own JS interop mechanism based on the <xref:Microsoft.JSInterop.IJSRuntime> interface, which is uniformly supported across Blazor hosting models and described in the following articles:

* <xref:blazor/js-interop/call-javascript-from-dotnet>
* <xref:blazor/js-interop/call-dotnet-from-javascript>

<xref:Microsoft.JSInterop.IJSRuntime> enables library authors to build JS interop libraries that can be shared across the Blazor ecosystem and remains the recommended approach for JS interop in Blazor.

> This article describes an alternative JS interop approach specific to WebAssembly-based apps available for the first time with the release of .NET 7. These approaches are appropriate when you only expect to run on client-side WebAssembly and not in the other Blazor hosting models. Library authors can use these approaches to optimize JS interop by checking at runtime if the app is running on WebAssembly in a browser (<xref:System.OperatingSystem.IsBrowser%2A?displayProperty=nameWithType>). The approaches described in this article should be used to replace obsolete unmarshalled JS interop API when migrating to .NET 7.

Reviewer 2: ```suggestion
This article describes an alternative JS interop approach specific to WebAssembly-based apps available for the first time with the release of .NET 7. These approaches are appropriate when you only expect to run on client-side WebAssembly and not in the other Blazor hosting models. Library authors can use these approaches to optimize JS interop by checking at runtime if the app is running on WebAssembly in a browser (<xref:System.OperatingSystem.IsBrowser%2A?displayProperty=nameWithType>). The approaches described in this article should be used to replace the obsolete unmarshalled JS interop API when migrating to .NET 7.
```

## Call JavaScript from .NET

This section explains how to call JS functions from .NET.

In the following `CallJavaScript1` component:

* The `CallJavaScript1` module is imported asynchronously from the [collocated JS file](xref:blazor/js-interop/index#load-a-script-from-an-external-javascript-file-js-collocated-with-a-component) with `JSHost.ImportAsync`.
* The imported `getMessage` JS function is called by `GetWelcomeMessage`.
* The returned welcome message string is displayed in the UI via the `message` field.

`Pages/CallJavaScript1.razor`:

```razor
@page "/call-javascript-1"
@using System.Runtime.InteropServices.JavaScript

<h1>
    .NET JS <code>[JSImport]</code>/<code>[JSExport]</code> Interop 
    (Call JS Example 1)
</h1>

@(message is not null ? message : string.Empty)

@code {
    private string? message;

    protected override async Task OnInitializedAsync()
    {
        await JSHost.ImportAsync("CallJavaScript1", 
            "../Pages/CallJavaScript1.razor.js");

        message = GetWelcomeMessage();
    }
}
```

> [!NOTE]

> > Code can include a conditional check for <xref:System.OperatingSystem.IsBrowser%2A?displayProperty=nameWithType> to ensure that the JS interop is only called in Blazor WebAssembly apps running on the client in a browser. This is important for libraries/NuGet packages that target Blazor WebAssembly and Blazor Server apps because Blazor Server apps can't execute the code provided by this JS interop API.

Reviewer 2: ```suggestion
> Code can include a conditional check for <xref:System.OperatingSystem.IsBrowser%2A?displayProperty=nameWithType> to ensure that the JS interop is only called in Blazor WebAssembly apps running on the client in a browser. This is important for libraries/NuGet packages that target Blazor hosting models that are not based on WebAssembly, like Blazor Server and Blazor Hybrid, which can't execute the code provided by this JS interop API.
```

    }
}
```

`Pages/CallDotNet2.razor`:

```razor
@page "/call-dotnet-2"
@using BlazorSample.JavaScriptInterop

<h1>
    .NET JS <code>[JSImport]</code>/<code>[JSExport]</code> Interop  
    (Call .NET Example 2)
</h1>

<p>
    <span id="result">.NET method not executed</span>
</p>

@code {
    protected override void OnAfterRender(bool firstRender)
    {
        Interop.SetWelcomeMessage();
    }
}
```

> [!IMPORTANT]
> In this section's example, JS interop is used to mutate a DOM element *purely for demonstration purposes* after the component is rendered in [`OnAfterRender`](xref:blazor/components/lifecycle#after-component-render-onafterrenderasync). Typically, you should only mutate the DOM with JS when the object doesn't interact with Blazor. The approach shown in this section is similar to cases where a third-party JS library is used in a Razor component, where the component interacts with the JS library via JS interop, the third-party JS library interacts with part of the DOM, and Blazor isn't involved directly with the DOM updates to that part of the DOM. For more information, see <xref:blazor/js-interop/index#interaction-with-the-document-object-model-dom>.

> ## Additional example
> 
> For an additional example of the JS interop techniques described in this article, see the *Blazor WASM demo of new JSInterop* sample app:
> 
> * [Reference source (`pavelsavara/blazor-wasm-hands-pose` GitHub repository)](https://github.com/pavelsavara/blazor-wasm-hands-pose)
> * [Live demonstration](https://pavelsavara.github.io/blazor-wasm-hands-pose/)
> 
> > [!NOTE]
> > The [`pavelsavara/blazor-wasm-hands-pose` GitHub repository](https://github.com/pavelsavara/blazor-wasm-hands-pose) isn't owned, maintained, or supported by the .NET foundation or Microsoft.
> >
> > The *Blazor WASM demo of new JSInterop* sample app uses a public JS library from [MediaPipe](https://www.mediapipe.dev/), which isn't owned, maintained, or supported by the .NET foundation or Microsoft.
> 

Reviewer 2: I'd rather not point to these samples in the docs as they are not maintained as part of the docs and could go stale.

```suggestion
```

Author: That's strange. I thought that I removed this on the commit.

## Additional resources

<!--

    HOLD: WAIT FOR CLIENT-SIDE ARTICLE TO MERGE
    
    * <xref:client-side/import-export-interop>

-->

> 
> * [Use .NET from any JavaScript app in .NET 7](https://devblogs.microsoft.com/dotnet/use-net-7-from-any-javascript-app-in-net-7/)

Reviewer 2: I prefer to avoid links to blog post content as the content is point-in-time and can quickly go stale.

```suggestion
```

To import a JS function to call it from C#, use the `[JSImport]` attribute on a C# method signature that matches the JS function's signature. The first parameter to the `[JSImport]` attribute is the name of the JS function to import, and the second parameter is the name of the [JS module](xref:blazor/js-interop/index#javascript-isolation-in-javascript-modules).

In the following example, `getMessage` is a JS function that returns a `string` for a module named `CallJavaScript1`. The C# method signature matches: No parameters are passed to the JS function, and the JS function returns a `string`. The JS function is called by `GetWelcomeMessage` in C# code.

`Pages/CallJavaScript1.razor.cs`:

```csharp
using System.Runtime.InteropServices.JavaScript;
using System.Runtime.Versioning;

namespace BlazorSample.Pages
{
    [SupportedOSPlatform("browser")]
    public partial class CallJavaScript1
    {
        [JSImport("getMessage", "CallJavaScript1")]
        internal static partial string GetWelcomeMessage();
    }
}
```

> The app's namespace for the preceding `CallJavaScript1` partial class is `BlazorServer`. The component's namespace is `BlazorSample.Pages`. If using the preceding component in a local test app, update the namespace to match the app. For example, the namespace is `ContosoApp.Pages` if the app's namespace is `ContosoApp`. For more information, see <xref:blazor/components/index#partial-class-support>.

Author: Yes ... just a typo. Thx.

The module name in the `[JSImport]` attribute and the call to load the module in the component, which is covered next with `JSHost.ImportAsync` syntax, should match. The module name must also be unique in the app. When authoring a library for deployment in a NuGet package, we recommend using the NuGet package namespace as a prefix in module names. In the following example, the module name reflects the `Contoso.InteropServices.JavaScript` package and a folder of user message interop classes (`UserMessages`):

```csharp
[JSImport("getMessage", 
    "Contoso.InteropServices.JavaScript.UserMessages.CallJavaScript1")]
```

If the JS function doesn't directly interact with the rendered Document Object Model (DOM), import the module in [`OnInitializedAsync`](xref:blazor/components/lifecycle#component-initialization-oninitializedasync). Call the imported JS function with the .NET interop method.

> If the JS module interacts directly with the component's rendered UI, import the module in [`OnAfterRenderAsync`](xref:blazor/components/lifecycle#after-component-render-onafterrenderasync) and call the imported JS function:

Author: hummmm 🤔 ... I think Steve implied that if ... for example ... some JS lib is imported **_by the library being used_** (perhaps from the Internet), that you wouldn't be guaranteed that its there that early. However, that's just a guess. I think we'd need to ask him about it.

Yes, tho ... I see what you mean here in this case ... in this particular example.

> If the JS module interacts directly with the component's rendered UI, import the module in [`OnAfterRenderAsync`](xref:blazor/components/lifecycle#after-component-render-onafterrenderasync) and call the imported JS function:

Author: No ... actually ... I think it might be this in the module of Steve's example ...

```javascript
mapboxgl.accessToken = '{ACCESS TOKEN}';
```

Doesn't the module need to load after render for that `accessToken` set to occur?

> If the JS module interacts directly with the component's rendered UI, import the module in [`OnAfterRenderAsync`](xref:blazor/components/lifecycle#after-component-render-onafterrenderasync) and call the imported JS function:

Reviewer 3: > Doesn't the module need to load after render for that accessToken set to occur?

I'm not an expert on ES6 modules, but I think the code in the module doesn't run until all of the modules it `import`s are loaded (so the `mapboxgl` object should be available by the time that line runs).

But either way, I don't think the component getting rendered would have any impact on the mechanics of the import procedure (with a possible exception being that the module initialization somehow requires the DOM to look a certain way, but this would be highly unusual).

The `addMapToElement()` invocation in Steve's example _does_ have to be called after the component is rendered, because it utilizes the DOM rendered by the component. I'm guessing the module import was also placed in `OnAfterRenderAsync()` because JS interop in `OnInitializedAsync()` won't work during Blazor Server prerendering. But that isn't a concern in the context of this new JS interop module, since it only works in WebAssembly anyway.

cc @SteveSandersonMS in case there's something I'm missing!

> If the JS module interacts directly with the component's rendered UI, import the module in [`OnAfterRenderAsync`](xref:blazor/components/lifecycle#after-component-render-onafterrenderasync) and call the imported JS function:

Author: :+1: ... I think the changes that I made on the commit resolve this aspect then. The only reference made to after-render now is for JS interop that mutates the DOM with the warning about the special cases when that's required (e.g., 3rd party lib).

```csharp
protected override async Task OnAfterRenderAsync(bool firstRender)
{
    if (OperatingSystem.IsBrowser() && firstRender)
    {
        await JSHost.ImportAsync("CallJavaScript1", "../Pages/CallJavaScript1.razor.js");

        ...
    }
}
```

If the code holds a reference to the <xref:System.Runtime.InteropServices.JavaScript.JSObject> returned by `JSHost.ImportAsync`, call <xref:System.Runtime.InteropServices.JavaScript.JSObject.Dispose%2A?displayProperty=nameWithType> when the component is disposed:

> 

Reviewer 3: Hm, I'm not sure it really makes sense to do this because module imports aren't scoped to the lifetime of Blazor components.

This example also kind of implies that _not_ holding a reference to the return value of `JSHost.ImportAsync` would be a memory leak, since no one can/will call `Dispose()` on it.

In the following example, a JS function named `setMessage` is imported from a collocated JS file.

The `setMessage` method:

* Calls `globalThis.getDotnetRuntime(0)` to expose the WebAssembly .NET runtime instance for calling exported .NET methods.
* Obtains the app assembly's JS exports. The name of the app's assembly in the following example is `BlazorSample`.
* Calls the `BlazorSample.Pages.CallDotNet1.GetMessageFromDotnet` method from the exports (`exports`). The returned value, which is the welcome message, is assigned to the `CallDotNet1` component's `<span>` text. The app's namespace is `BlazorSample`, and the `CallDotNet1` component's namespace is `BlazorSample.Pages`.

`Pages/CallDotNet1.razor.js`:

```javascript
export async function setMessage() {
  const { getAssemblyExports } = await globalThis.getDotnetRuntime(0);

>   var exports = await getAssemblyExports("BlazorSample.dll");

Reviewer 3: In real world apps, you could probably put these two lines in a JS initializer (or modify the logic here to cache `exports`), since they only have to be executed once. But it might be fine to leave as-is in this example, for the sake of simplicity.

## Use of a single JavaScript module across components

*Before following the guidance in this section, read the [Call JavaScript from .NET](#call-javascript-from-net) and [Call .NET from JavaScript](#call-net-from-javascript) sections of this article, which provide general guidance on .NET JS `[JSImport]`/`[JSExport]` interop.*

The example in this section shows how to use JS interop from a shared JS module. The following components, classes, C# methods, and JS functions are used:

* `Interop` class (`Interop.cs`): Sets up import and export JS interop with the `[JSImport]` and `[JSExport]` attributes for a module named `Interop`.
  * `GetWelcomeMessage`: .NET method that calls the imported `getMessage` JS function.
  * `SetWelcomeMessage`: .NET method that calls the imported `setMessage` JS function.
  * `GetMessageFromDotnet`: An exported C# method that returns a welcome message string when called from JS.
* `wwwroot/js/interop.js` file: Contains the JS functions.
  * `getMessage`: Returns a welcome message when called by C# code in a component.
  * `setMessage`: Calls the `GetMessageFromDotnet` C# method and assigns the returned welcome message to a DOM `<span>` element.

> * `App` component (`App.razor`): Loads the module from `wwwroot/js/interop.js` by calling `JSHost.ImportAsync` in the [`OnInitializedAsync` event](xref:blazor/components/lifecycle#component-initialization-oninitializedasync).

Reviewer 3: I believe the module loading can go in `Program.cs`.

Author: Yes, that works. Are we going to show both, or is only ever doing it in `Program.cs` the most sensible approach?

Author: Another question is where this would be done for a library?

Reviewer 3: > ... and I now guess that he does that because of this var set in the module ...

I mentioned this in my other comment, but I believe this is actually because JS interop won't work in `OnInitializedAsync()` during Blazor Server prerendering (which isn't a concern for this new JS interop API, which is Wasm-only).

> Yes, that works. Are we going to show both, or is only ever doing it in `Program.cs` the most sensible approach?

If you're writing an RCL, `Program.cs` is definitely not an option. I think the best choice in the library case is to place module loading in component initialization (as is done in most of your examples). A possible downside is that could be slightly confusing because it gives the illusion that lifetime of the loaded module is somehow related to lifetime of the component (as it might be for modules imported as an `IJSObjectReference`). It would be cool if there was a better/easier/more automatic way to load JS modules using this new API, but I think that would have to come in the form of an additional framework feature.

But if you're going to place module loading in `App.razor`, you're probably not writing an RCL, so the logic can just go in `Program.cs`. IMO, that's slightly cleaner because the `App` component doesn't really have anything to do with module loading, and `Program.cs` is already used for app-wide configuration.

cc @SteveSandersonMS again in case you have any objections to any of this 😄

Author: I'll call out only `Program.cs` for the example. I'll indicate that this particular scenario ... the 3rd section approach ... is best applied to apps, not RCLs.

}
```

> [!NOTE]

> > Calling `globalThis.getDotnetRuntime(0)` to obtain the exports can occur in a [JavaScript initializer](xref:blazor/js-interop/index#javascript-initializers) for availability across the app.

Reviewer 3: This kind of makes it sound like `getDotnetRuntime(0)` is obtaining the exports, but that's what `getAssemblyExports` is doing. Maybe changing it to...
```suggestion
> Calling `getAssemblyExports` to obtain the exports can occur in a [JavaScript initializer](xref:blazor/js-interop/index#javascript-initializers) for availability across the app.
```
...is sufficient?

}
```

There are two options for loading the module:

* In `Program.cs`:

  ```csharp
  using System.Runtime.InteropServices.JavaScript;

  ...

>   await JSHost.ImportAsync("Interop", "../js/interop.js");

Reviewer 3: Maybe add the `IsBrowser()` check here? Your call - maybe it's already clear enough from context that this would be necessary.

Reviewer 3: Actually, thinking about this more, I'm wondering if it's really necessary to explicitly to include the `IsBrowser()` checks in these examples at all. It would be a fairly advanced scenario to optimize a library to use these new JS interop APIs on WebAssembly only, and fall back on `IJSRuntime` when on Blazor server. I would think that in most cases, someone using these new JS interop APIs is already assuming that they're in a WASM context. We could still mention performing an `IsBrowser()` check as a way to check if you're running in a browser in case you're writing components meant to work for both Blazor Server and Blazor WebAssembly.

Just a thought - feel free to ignore it if you don't agree!

Reviewer 3: Nevermind - it looks like you've already done this! 😄

Author: I agree. I just yanked them from the last example ... I'll go pull them from the code above and then just leave a NOTE about it in the text.
